src/core/ipns/republisher.js

```diff
@@ -0,0 +1,161 @@
+'use strict'
+
+const ipns = require('ipns')
+const crypto = require('libp2p-crypto')
+const PeerId = require('peer-id')
+const errcode = require('err-code')
+
+const debug = require('debug')
+const each = require('async/each')
+const waterfall = require('async/waterfall')
+const log = debug('jsipfs:ipns:republisher')
+log.error = debug('jsipfs:ipns:republisher:error')
+
+const minute = 60 * 1000
+const hour = 60 * minute
+
+const defaultBroadcastInterval = 4 * hour
+const defaultRecordLifetime = 24 * hour
+
+class IpnsRepublisher {
+  constructor (publisher, ipfs) {
+    this._publisher = publisher
+    this._ipfs = ipfs
+    this._repo = ipfs._repo
+    this._timeoutId = null
+    this._canceled = false
+    this._onCancel = null
+  }
+
+  start () {
+    const periodically = (cb) => {
+      this._republishEntries(this._ipfs._peerInfo.id.privKey, this._ipfs._options.pass, () => {
+        if (this._canceled) {
+          return this._onCancel()
+        }
+        this._timeoutId = setTimeout(() => periodically(cb), defaultBroadcastInterval)
+      })
+    }
+
+    setTimeout(() => {
+      periodically()
+    }, minute)
+  }
+
+  stop (cb) {
+    this._canceled = true
+    if (this._timeoutId || !this._onCancel) {
+      // Not running
+      clearTimeout(this._timeoutId)
+      return cb()
+    }
+
+    this._onCancel = cb
+  }
```

Comment:
I think this needs a bit of a refactor so that start and stop can be called multiple times. Something like this should work:

```js
  constructor (publisher, ipfs) {
    this._publisher = publisher
    this._ipfs = ipfs
    this._repo = ipfs._repo
    this._republishHandle = null
  }

  start () {
    if (this._republishHandle) throw new Error('already running')

    const republishHandle = {
      _onCancel: null,
      _timeoutId: null,
      runPeriodically (fn, period) {
        republishHandle._timeoutId = setTimeout(() => {
          republishHandle._timeoutId = null

          fn((nextPeriod) => {
            // Was republish cancelled while fn was being called?
            if (republishHandle._onCancel) {
              return republishHandle._onCancel()
            }
            // Schedule next
            republishHandle.schedule(fn, nextPeriod || period)
          })
        }, period)
      },
      cancel (cb) {
        // Not currently running a republish, can callback immediately
        if (republishHandle._timeoutId) {
          clearTimeout(republishHandle._timeoutId)
          return cb()
        }
        // Wait for republish to finish then call callback
        republishHandle._onCancel = cb
      }
    }

    const { privKey } = this._ipfs._peerInfo.id
    const { pass } = this._ipfs._options

    republishHandle.runPeriodically((done) => {
      this._republishEntries(privKey, pass, () => done(defaultBroadcastInterval))
    }, minute)

    this._republishHandle = republishHandle
  }

  stop (cb) {
    const republishHandle = this._republishHandle

    if (!republishHandle) {
      return cb(new Error('not running'))
    }

    this._republishHandle = null
    republishHandle.cancel(cb)
  }
```

Reply:
you are right Alan! Thanks 🙂